ios/engine/KMEI/KeymanEngine/Classes/Resource Management/AssociatingPackageInstaller.swift

```diff
@@ -0,0 +1,380 @@
+
+    // Specifications needed across multiple threads.
+    let associationSpecs: [Associator]
+    var associationInstallers: [Int: AssociationInstaller] = [:]
+
+    // Holds data needed for progress tracking.
+    var pickingCompleted: Bool = false
+    var associationQueryProgress: [Int: LanguagePickAssociator.Progress] = [:]
+    var installProgressMap: [KeymanPackage.Key: PackageInstallResult?] = [:]
+
+    let progressCallback: ProgressReceiver
+    let downloadManager: ResourceDownloadManager
+
+    var isCancelled = false
+
+    init(with associationSpecs: [Associator], downloadManager: ResourceDownloadManager, receiver: @escaping ProgressReceiver) {
+      self.associationSpecs = associationSpecs
+      self.downloadManager = downloadManager
+      self.progressCallback = receiver
+    }
+
+    // Since progress info is stored here, it makes the most sense to track progress-related
+    // methods here too.
+
+    internal var reportsProgress: Bool {
+      return !isCancelled && pickingCompleted
+    }
+
+    /**
+     * Computes the initial level of progress made toward the overall installation at the time that language selections were
+     * finalized.
+     */
+    internal func initializeProgress() {
+      self.progressCallback(.starting)
+    }
+
+    /**
+     * Generates data useful for reporting about the installation's overall progress.
+     */
+    internal func reportProgress(complete: Bool = false) {
+      if reportsProgress {
+        if complete {
+          progressCallback(.complete)
+        } else {
+          progressCallback(.inProgress)
+        }
+      }
+    }
+
+    internal func reportCancelled() {
+      progressCallback(.cancelled)
+    }
+  }
+
+  private let closureShared: ClosureShared
+
+  let package: Package
+  let defaultLgCode: String?
+  var associationQueriers: [LanguagePickAssociator]?
+
+  public convenience init(for package: Package,
+                          defaultLanguageCode: String? = nil,
+                          withAssociators associators: [Associator] = [],
+                          progressReceiver: @escaping ProgressReceiver) {
+    self.init(for: package,
+              defaultLanguageCode: defaultLanguageCode,
+              downloadManager: ResourceDownloadManager.shared,
+              withAssociators: associators,
+              progressReceiver: progressReceiver)
+  }
+
+  internal init(for package: Package,
+                defaultLanguageCode: String? = nil,
+                downloadManager: ResourceDownloadManager = ResourceDownloadManager.shared,
+                withAssociators associators: [Associator] = [],
+                progressReceiver: @escaping ProgressReceiver) {
+    self.package = package
+    self.defaultLgCode = defaultLanguageCode
+
+    self.closureShared = ClosureShared(with: associators, downloadManager: downloadManager, receiver: progressReceiver)
+  }
+
+  // TODO:  deinit handling in case everything gets cancelled.
```

Comment:
Is this coming in a future PR?

Reply:
Definitely should... though probably as an issue first, since there are a few higher-priority aspects to worry about right now.

Reply:
Now addressed by #3468.